cpp/src/text/subword/detail/tokenizer_utils.cuh

```diff
@@ -16,39 +16,61 @@
 
 #pragma once
 
+#include <text/subword/detail/cp_data.h>
+
 #include <stdint.h>
-#include <vector>
 
 namespace nvtext {
 namespace detail {
 
 constexpr int THREADS_PER_BLOCK = 64;
 
 /**
- * @brief A selection op for cub to get elements from an array not equal to a certain value.
+ * @brief In-place update of offsets values.
  *
- * See https://nvlabs.github.io/cub/structcub_1_1_device_partition.html for an example of
- * this struct.
- */
-struct NotEqual {
-  uint32_t const val_to_omit;
-
-  __host__ __device__ NotEqual(uint32_t const val_to_omit) : val_to_omit(val_to_omit) {}
-
-  __host__ __device__ bool operator()(uint32_t const& a) const { return (a != val_to_omit); }
-};
-
-/**
- * @brief In-place update of lengths values.
+ * In the `d_chars_up_to_idx`, the last character of each string is basically
+ * the offset (i.e. the number of characters) in that string.
+ *
+ * Example
+ * @code{.pseudo}
+ * // 3 strings with sizes 5,4,2
+ * d_offsets = [0,5,9,11]
+ * // code points generated per character (as offsets)
+ * // 2nd string has an extra code point at its first char
+ * d_chars_up_to_idx = [1,2,3,4,5,6,8,9,10,11,12]
+ * d_chars_up_to_idx[d_offsets[1-3]] is [5,10,12]
+ * => d_offsets becomes [0,5,10,12]
+ * @endcode
  */
 struct update_strings_lengths_fn {
   uint32_t const* d_chars_up_to_idx;
-  uint32_t* d_lengths;
+  uint32_t* d_offsets;
   __device__ void operator()(uint32_t idx)
   {
-    d_lengths[idx] = d_chars_up_to_idx[d_lengths[idx] - 1];
+    auto const offset = d_offsets[idx];
+    d_offsets[idx]    = offset > 0 ? d_chars_up_to_idx[offset - 1] : 0;
   }
 };
 
+/**
+ * @brief Retrieve the code point metadata table.
+ *
+ * This is a singleton instance that copies a large table of integers into
+ * device memory on the very first call.
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ */
+codepoint_metadata_type const* get_codepoint_metadata(cudaStream_t stream);
```

Comment:
How large? Any need to provide functions to unload and free up the memory?

Reply:
The two tables are 4456KB (1114112 x 4) and 954KB (119233 x 8) so together only 5.4MB.